README.md
# flutter_barometer_plugin

This plugin makes it easier to use barometer of devices.

## Getting Started

Use this package as a library
Depend on it
Run this command:

With Flutter:

` $ flutter pub add flutter_barometer_plugin`

This will add a line like this to your package's pubspec.yaml (and run an implicit dart pub get):


``` yaml
dependencies: 
  flutter_barometer_plugin: ^0.0.1+3
```

Alternatively, your editor might support flutter pub get. Check the docs for your editor to learn more.

Import it
Now in your Dart code, you can use:

`import 'package:flutter_barometer_plugin/flutter_barometer.dart';`

## Usage 

### Example

``` dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_barometer/flutter_barometer.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  BarometerValue _currentPressure = BarometerValue(0.0);

  @override
  void initState() {
    super.initState();
    FlutterBarometer.currentPressueEvent.listen((event) {
      setState(() {
        _currentPressure = event;
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('flutter_barometer'),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Text(
                '${(_currentPressure.hectpascal * 1000).round() / 1000} hPa',
                style: TextStyle(
                  fontSize: 70,
                ),
              ),
              Text(
                '${(_currentPressure.inchOfMercury * 1000).round() / 1000} inHg',
                style: TextStyle(
                  fontSize: 30,
                ),
              ),
              Text(
                '${(_currentPressure.millimeterOfMercury * 1000).round() / 1000} mmHg',
                style: TextStyle(
                  fontSize: 30,
                ),
              ),
              Text(
                '${(_currentPressure.poundsSquareInch * 1000).round() / 1000} psi',
                style: TextStyle(
                  fontSize: 30,
                ),
              ),
              Text(
                '${(_currentPressure.atm * 1000).round() / 1000} atm',
                style: TextStyle(
                  fontSize: 30,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```




CHANGELOG.md
## 0.0.1+3
add unit
## 0.0.1+2
renew Readme
## 0.0.1+1
add license

## 0.0.1
 initial release.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_barometer_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/flutter_barometer_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_barometer/flutter_barometer.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  const MethodChannel channel = MethodChannel('flutter_barometer');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await FlutterBarometer.platformVersion, '42');
  });
}


lib/flutter_barometer.dart
import 'dart:async';

import 'package:flutter/services.dart';

class FlutterBarometer {
  static const MethodChannel _channel =
      const MethodChannel('flutter_barometer');
  static const EventChannel _barometerEventChannel =
      EventChannel('pressureStream');

  static var _onPressureChanged;

  static Future<String?> get platformVersion async {
    final String? version = await _channel.invokeMethod('getPlatformVersion');
    return version;
  }

  static Future<BarometerValue> get currentPressure async {
    final pressure = await _channel.invokeMethod('getCurrentPressure');
    return BarometerValue(pressure);
  }

  static Stream<BarometerValue> get currentPressureEvent {
    if (_onPressureChanged == null) {
      _onPressureChanged = _barometerEventChannel.receiveBroadcastStream().map(
            (element) => BarometerValue(element as double),
          );
    }
    return _onPressureChanged;
  }

}

class BarometerValue {
  const BarometerValue(this.hectpascal);
  final double hectpascal;

  double get inchOfMercury => hectpascal / 33.8639;
  double get millimeterOfMercury => hectpascal / 1.33322;
  double get poundsSquareInch => hectpascal / 68.9476;
  double get atm => hectpascal / 1013.25;
}


ios/Classes/SwiftFlutterBarometerPlugin.swift
import Flutter
import UIKit
import CoreMotion

public class SwiftFlutterBarometerPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    var altimeter:CMAltimeter?
    var pressure:Double = 0.0
    var sinkOnChanged: FlutterEventSink?
    var listeningOnChanged = false
    
    override init() {
        super.init()
        altimeter = CMAltimeter()
        getCurrentPressure()
    }
    
    private func getCurrentPressure() {
        if(CMAltimeter.isRelativeAltitudeAvailable()) {
            altimeter!.startRelativeAltitudeUpdates(
                to: OperationQueue.main,
                withHandler:
                    {data, error in
                        if error == nil {
                            let pressure:Double = data!.pressure.doubleValue
                            //let altitude:Double = data!.relativeAltitude.doubleValue
                            self.pressure =  pressure * 10
                            self.onChanged(value: self.pressure);
                        }
                    }
            )
        }
    }
    
    //まずこれが呼ばれる
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "flutter_barometer", binaryMessenger: registrar.messenger())
        let stream = FlutterEventChannel(name: "pressureStream", binaryMessenger: registrar.messenger())
        let instance = SwiftFlutterBarometerPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        stream.setStreamHandler(instance)
    }
    
    // EventChannel 初期化
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.sinkOnChanged  = events
        listeningOnChanged = true
        return nil
    }
    
    // EventChannel 終了処理
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        listeningOnChanged = false
        return nil
    }
    
    // EventChannel 値を流す
    public func onChanged(value: Double) {
        if listeningOnChanged {
            if let sink = self.sinkOnChanged {
                sink(value)
            }
        }
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        // これが呼ばれた時に
        if call.method == "getPlatformVersion" {
            // これを返す
            result("iOS " + UIDevice.current.systemVersion)
        }
        
        if call.method == "getCurrentPressure" {
            result(self.pressure)
        }
    }
}


android/src/main/java/com/example/flutter_barometer/FlutterBarometerPlugin.java
package com.example.flutter_barometer;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

/**
 * FlutterBarometerPlugin
 */
public class FlutterBarometerPlugin implements FlutterPlugin, MethodCallHandler, SensorEventListener {
    private MethodChannel channel;
    private EventChannel stream;
    private SensorManager sensorManager;
    private EventChannel.EventSink sinkOnChanged;
    public Sensor pressureSensor;
    private Context applicationContext = null;

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        applicationContext = null;
        sensorManager.unregisterListener(this);
        channel.setMethodCallHandler(null);
    }

    @Override
    public void onSensorChanged(SensorEvent sensorEvent) {
        double pressure = sensorEvent.values[0];
        if (sinkOnChanged != null) {
            sinkOnChanged.success(pressure);
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int i) {

    }

    private boolean initializePressureSensor() {
        sensorManager = (SensorManager) applicationContext.getSystemService(Context.SENSOR_SERVICE);
        if (sensorManager.getDefaultSensor(Sensor.TYPE_PRESSURE) != null) {
            pressureSensor = sensorManager.getDefaultSensor(Sensor.TYPE_PRESSURE);
            sensorManager.registerListener(this, pressureSensor, SensorManager.SENSOR_DELAY_NORMAL);
            return true;
        } else {
            pressureSensor = null;
            return false;
        }
    }

    private EventChannel.StreamHandler handler() {
        return new EventChannel.StreamHandler() {
            @Override
            public void onListen(Object listener, EventChannel.EventSink events) {
                if (initializePressureSensor()) {
                    sinkOnChanged = events;
                }
            }

            @Override
            public void onCancel(Object args) {
                sinkOnChanged = null;
            }
        };
    }

    @Override
    public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
        applicationContext = flutterPluginBinding.getApplicationContext();
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "flutter_barometer");
        channel.setMethodCallHandler(this);
        stream = new EventChannel(flutterPluginBinding.getBinaryMessenger(), "pressureStream");
        stream.setStreamHandler(handler());
    }

    // こいつが呼ばれる
    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        // getPlatFormVersionが呼ばれた時
        if (call.method.equals("getPlatformVersion")) {
            // この値が返る
            result.success("Android " + android.os.Build.VERSION.RELEASE);
            return;
        }
        result.notImplemented();
    }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_barometer/flutter_barometer.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';
  BarometerValue _currentPressure = BarometerValue(0.0);

  @override
  void initState() {
    super.initState();
    initPlatformState();
    FlutterBarometer.currentPressureEvent.listen((event) {
      setState(() {
        _currentPressure = event;
      });
    });
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    String platformVersion;
    // Platform messages may fail, so we use a try/catch PlatformException.
    // We also handle the message potentially returning null.
    try {
      platformVersion =
          await FlutterBarometer.platformVersion ?? 'Unknown platform version';
    } on PlatformException {
      platformVersion = 'Failed to get platform version.';
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _platformVersion = platformVersion;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('flutter_barometer'),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Text(
                'Running on: $_platformVersion\n',
                style: TextStyle(fontSize: 30),
              ),
              Text(
                '${(_currentPressure.hectpascal * 1000).round() / 1000} hPa',
                style: TextStyle(
                  fontSize: 70,
                ),
              ),
              Text(
                '${(_currentPressure.inchOfMercury * 1000).round() / 1000} inHg',
                style: TextStyle(
                  fontSize: 30,
                ),
              ),
              Text(
                '${(_currentPressure.millimeterOfMercury * 1000).round() / 1000} mmHg',
                style: TextStyle(
                  fontSize: 30,
                ),
              ),
              Text(
                '${(_currentPressure.poundsSquareInch * 1000).round() / 1000} psi',
                style: TextStyle(
                  fontSize: 30,
                ),
              ),
              Text(
                '${(_currentPressure.atm * 1000).round() / 1000} atm',
                style: TextStyle(
                  fontSize: 30,
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
